x/payment/types/payfordata.go

```diff
@@ -155,16 +155,13 @@ func CreateCommitment(k uint64, namespace, message []byte) ([]byte, error) {
 	// create the commits by pushing each leaf set onto an nmt
 	subTreeRoots := make([][]byte, len(leafSets))
 	for i, set := range leafSets {
-		// create the nmt todo(evan) use nmt wrapper
-		tree := nmt.New(sha256.New(), nmt.NamespaceIDSize(NamespaceIDSize))
+		tree := wrapper.NewErasuredNamespacedMerkleTree(consts.MaxSquareSize)
 		for _, leaf := range set {
 			nsLeaf := append(make([]byte, 0), append(namespace, leaf...)...)
-			err := tree.Push(nsLeaf)
-			if err != nil {
-				return nil, err
-			}
+			// here we hardcode pushing as axis 0 cell 0 because we never want
+			// to add the parity namespace to our shares when we create roots.
+			tree.Push(nsLeaf, rsmt2d.SquareIndex{Axis: 0, Cell: 0})
```

Comment:
[question] it's not clear to me how this is related to https://github.com/celestiaorg/celestia-core/blob/3c6122ae98f3ea98d4885a986d3e8d5ec1077d45/pkg/consts/consts.go#L69

Can you please elaborate?

Reply:
yeah, so we have to ensure that the namespace is included in erasure data. To do that, we have to include the namespace in the share. This means that the erasure data will not have a namspace, so efffectively this wrapper adds the correct namespace to the data so that nmt knows how to interpret it. 

here we add the parity namespace to a share, depending on its position
https://github.com/celestiaorg/celestia-core/blob/08c0c038e63a7848f2761beb1013ee102883be98/pkg/wrapper/nmt_wrapper.go#L58

Reply:
by hardcoding 0,0, we're forcing it to use the namespace we want (the namespace of the share) instead of potentially going over the square size and then accidentally using an incorrect namespace.

Reply:
thanks for explaining! It seems like this line is tightly coupled to an implementation detail of https://github.com/celestiaorg/celestia-core/blob/08c0c038e63a7848f2761beb1013ee102883be98/pkg/wrapper/nmt_wrapper.go#L50

To make this more explicit, I wonder if we may add a method in a subsequent refactor:

```golang
// PushWithNamespace adds the provided data to the underlying
// NamespaceMerkleTree and preserves the first DefaultNamespaceIDLen bytes as
// the namespace for this data. Fulfills the rsmt.Tree interface. NOTE: panics
// if an error is encountered while pushing or if the tree size is exceeded.
func (w *ErasuredNamespacedMerkleTree) PushWithNamespace(data []byte) {
	nidAndData := make([]byte, consts.NamespaceSize+len(data))
	copy(nidAndData[consts.NamespaceSize:], data)
	copy(nidAndData[:consts.NamespaceSize], data[:consts.NamespaceSize])
	// push to the underlying tree
	err := w.tree.Push(nidAndData)
	// panic on error
	if err != nil {
		panic(err)
	}
}
```

Reply:
yeah, I like this idea. I'd also be down for refactoring the wrapped nmt into something else entirely. It relies on rsmt2d to call it, which we don't actually have to do. During nmt cacheing to check for message inclusion, this makes things even more confusing and suboptimal as we have to make assumtions over when and how rsmt2d actually calls the tree and generates commitments.

a while ago we considered not passing some nmt (or wrapped nmt) object to rsmt, and instead simply reading the data we need as we need it. https://github.com/celestiaorg/rsmt2d/pull/9#issuecomment-736805729 . We don't actually need to change anything in rsmt2d for this, we can just code it up.